Add reset_grad() function (#42754)

ngimel opened the pull request:
Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/42754

Test Plan: Imported from OSS

Differential Revision: D23010859

Pulled By: firstprayer

mcarilli commented on torch/nn/modules/module.py:
```diff
@@ -1313,8 +1313,14 @@ def requires_grad_(self: T, requires_grad: bool = True) -> T:
-        r"""Sets gradients of all model parameters to zero."""
+    def zero_grad(self, set_to_none: bool = False) -> None:
+        r"""Sets gradients of all model parameters to zero. See similar function
+        under `torch.optimizer` for more contexts.
```
```
``torch.optimizer``
```

mcarilli commented on torch/optim/optimizer.py:
```diff
@@ -164,16 +164,32 @@ def update_group(group, new_group):
+                However, it changes certain behaviors. For example:
+                1. When user tries to access the gradient value and perform manual ops on it.
+                A None attribute or a Tensor full of 0s will be different.
+                2. If the user requests `zero_grad(set_to_none=True)` followed by a backward pass, `.grad` s
```
```
``zero_grad(set_to_none=True)`` followed by a backward pass, ``.grad``\ s
```

mcarilli commented on torch/optim/optimizer.py:
```diff
@@ -164,16 +164,32 @@ def update_group(group, new_group):
+                A None attribute or a Tensor full of 0s will be different.
+                2. If the user requests `zero_grad(set_to_none=True)` followed by a backward pass, `.grad` s
+                are guaranteed to be None for params that did not receive a gradient.
+                3. `torch.optim` optimizers have a different behavior if the gradient is 0 or None
```
```
``torch.optim``
```

Commit:
Add reset_grad() function (#44423)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/44423

Pull Request resolved: https://github.com/pytorch/pytorch/pull/42754

Test Plan: Imported from OSS

Reviewed By: mruberry

Differential Revision: D23010859

Pulled By: ngimel

fbshipit-source-id: 760279f7c9cb84d11bef51207c18bf1f362ca7ad
